gtk: apply all window appearance changes in syncAppearance

pluiedev opened the pull request:
The GTK side of appearance code is kind of a mess with several different functions all having the responsibility of interacting with each other and setting the appropriate window appearance. It should solely be the responsibility of the `syncAppearance` function to apply appearance changes, with other callbacks/functions calling it instead: much like what we already do for the macOS apprt.

~~I also took the time to refactor the libadwaita version checks since calling `versionAtLeast(0, 0, 0)` does get old after a while. Now almost all checks are given human-readable names and contributors need not memorize what the relevant version checks all are.~~ Moved to another PR

pluiedev committed:
gtk: apply all window appearance changes in syncAppearance

The GTK side of appearance code is kind of a mess with several different
functions all having the responsibility of interacting with each other
and setting the appropriate window appearance. It should solely be the
responsibility of the `syncAppearance` function to apply appearance
changes, with other callbacks/functions calling it instead: much like
what we already do for the macOS apprt.

mitchellh commented:
Just a note that just to be extra safe I'm going to push this after the 1.1 release.

pluiedev commented:
> Just a note that just to be extra safe I'm going to push this after the 1.1 release.

On Discord I've seen a bunch of bugs now that can be fixed by calling syncAppearance more regularly and in more places, and I think it'd be great if this can be part of 1.1.1. I can move the Adwaita refactors to another PR.

pluiedev committed:
config: update reload ability information for several keys

Several keys are now able to affect existing windows (especially
window-decoration, whose config documentation got a greater overhaul)

jcollie commented:
@pluiedev if you can rebase this on the current main I'd love to review this PR this evening.

jcollie commented on src/apprt/gtk/Window.zig:
```diff
@@ -328,53 +355,89 @@ pub fn updateConfig(
+        // Disable the title buttons (close, maximize, minimize, ...)
+        // *inside* the tab overview if CSDs are disabled.
+        // We do spare the search button, though.
+        c.adw_tab_overview_set_show_start_title_buttons(tab_overview, @intFromBool(csd_enabled));
```
Not compiling right now because of missing `@ptrCast`s

jcollie commented on src/apprt/gtk/Window.zig:
```diff
@@ -55,6 +57,38 @@ adw_tab_overview_focus_timer: ?c.guint = null,
 /// State and logic for windowing protocol for a window.
 winproto: winproto.Window,
 
+pub const DerivedConfig = struct {
```
Yay for DerivedConfig!

jcollie commented on src/apprt/gtk/Window.zig:
```diff
@@ -328,53 +355,89 @@ pub fn updateConfig(
-        // changes necessary so we just log the error and continue.
-        log.warn("failed to update window protocol config error={}", .{err});
-    };
+    self.config = DerivedConfig.init(config);
```
Copying over is fine for now, but we may need to make this a little more sophisticated if we ever need to alloc parts of the derived config.
